out variant for cudnn batch norm

makslevental opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #61824
* #61821
* #61802
* #61801
* #61800

Differential Revision: [D29812546](https://our.internmc.facebook.com/intern/diff/D29812546)

makslevental committed:
out variant for cudnn batch norm

[ghstack-poisoned]

makslevental committed:
Update on "out variant for cudnn batch norm"




[ghstack-poisoned]

Commit:
Update on "out variant for cudnn batch norm"




[ghstack-poisoned]

Commit:
Update on "out variant for cudnn batch norm"




[ghstack-poisoned]

Commit:
Update on "out variant for cudnn batch norm"




[ghstack-poisoned]

makslevental commented:
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29812546).

makslevental committed:
Update on "out variant for cudnn batch norm"


Differential Revision: [D29812546](https://our.internmc.facebook.com/intern/diff/D29812546)

[ghstack-poisoned]

makslevental committed:
Update on "out variant for cudnn batch norm"


Differential Revision: [D29812546](https://our.internmc.facebook.com/intern/diff/D29812546)

[ghstack-poisoned]

desertfire commented on aten/src/ATen/native/cudnn/BatchNorm.cpp:
```diff
@@ -195,41 +201,103 @@ std::tuple<Tensor, Tensor, Tensor, Tensor> cudnn_batch_norm(
+    save_mean = at::empty({num_features}, weight_t.options());
+    save_var = at::empty({num_features}, weight_t.options());
+
+#if CUDNN_VERSION >= 7400
```
`reserve` is not allocated for `#else` or `!training`?

makslevental replied:
my reasoning here was three fold:

1) a reserve buffer is indeed not used, neither for cudnn < 7.4 nor in forward pass  
2) up above we have this

>  // save_mean and save_var can be undefined
  // If this causes problems, we can initialize them to empty tensors
  // of the correct type
  
 thus reasoning by analogy...

3) return is ultimately consumed in [Tensor batch_norm](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Normalization.cpp#L538)
```
return std::get<0>(at::_batch_norm_impl_index(input, weight, bias, running_mean, running_var, training, momentum, eps, cudnn_enabled));
```
which discards. 

but it was just my intuition 🤷. if you think this would break something i will certainly put it back.

desertfire commented on aten/src/ATen/native/cudnn/BatchNorm.cpp:
```diff
@@ -195,41 +201,103 @@ std::tuple<Tensor, Tensor, Tensor, Tensor> cudnn_batch_norm(
+#endif // CUDNN_VERSION >= 7400
+  }
+
+  std::tie(output_t, save_mean, save_var, reserve) = cudnn_batch_norm_out(
```
Would linter complain if we skip the assignment to std::tie?

makslevental replied:
C++ doesn't do this kind of unpacking before [`std=c++17`](https://en.cppreference.com/w/cpp/language/structured_binding) and our cmake sets [`std=c++14`](https://github.com/pytorch/pytorch/blob/master/CMakeLists.txt#L40)

makslevental replied:
oh duh whoops. will change. thanks!

desertfire commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1309,6 +1309,10 @@
   dispatch:
     CUDA: cudnn_batch_norm
 
+- func: cudnn_batch_norm.out(Tensor input, Tensor weight, Tensor? bias, Tensor? running_mean, Tensor? running_var, bool training, float exponential_average_factor, float epsilon, *, Tensor(a!) out, Tensor(b!) mean, Tensor(c!) var, Tensor(d!) reserve) -> (Tensor(a!), Tensor(b!), Tensor(c!), Tensor(d!))
```
Nit: keep the new parameter naming consistent? e.g. `save_mean`, `save_var`

makslevental committed:
Update on "out variant for cudnn batch norm"


Differential Revision: [D29812546](https://our.internmc.facebook.com/intern/diff/D29812546)

[ghstack-poisoned]

makslevental committed:
Update on "out variant for cudnn batch norm"


Differential Revision: [D29812546](https://our.internmc.facebook.com/intern/diff/D29812546)

[ghstack-poisoned]

makslevental commented:
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29812546).

makslevental commented:
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29812546).

vadimkantorov commented:
This does not modify for now the user-facing nn.BatchNorm*d and F.batch_norm, right?

makslevental committed:
Update on "out variant for cudnn batch norm"


Differential Revision: [D29812546](https://our.internmc.facebook.com/intern/diff/D29812546)

[ghstack-poisoned]

makslevental committed:
Update on "out variant for cudnn batch norm"


Differential Revision: [D29812546](https://our.internmc.facebook.com/intern/diff/D29812546)

[ghstack-poisoned]

makslevental committed:
Update on "out variant for cudnn batch norm"


Differential Revision: [D29812546](https://our.internmc.facebook.com/intern/diff/D29812546)

[ghstack-poisoned]

vadimkantorov commented:
@ezyang if the CLA is signed, maybe staleness could be reevaluated?
